### PR TITLE
Add Linux support

### PR DESCRIPTION
**Pull request by yassernasc**
Calculation reference: https://stackoverflow.com/questions/16726779/how-do-i-get-the-total-cpu-usage-of-an-application-from-proc-pid-stat

Info parsing:
https://www.man7.org/linux/man-pages/man5/proc_uptime.5.html (get the whole system uptime)

https://www.man7.org/linux/man-pages/man5/proc_pid_task.5.html (get process child threads)

https://www.man7.org/linux/man-pages/man5/proc_pid_stat.5.html (get individual thread stats)

### CODE REVIEW

File: linux.c [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+#include <unistd.h>
+
+#include "types.h"
+
+static inline int
+thread_stats__linux (thread_stats_t *stats, size_t *len) {
+  int self = getpid();
+  clock_t clock_ticks = sysconf(_SC_CLK_TCK);
+
+  *len = 0;
```
Comment (kasperisager) [inline linux.c]: The initial value of `*len` is the size of `stats`, which we must ensure doesn't overflow in the loop further down. Track the number of written elements separately and only _after_ the loop assign this value to `*len`.

File: binding.c [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 #if THREAD_STATS_APPLE
 #include "apple.h"
+#elif THREAD_STATS_LINUX
+#include "linux.c"
```
Comment (kasperisager) [inline binding.c]:
```suggestion
#include "linux.h"
```

Only header files should be `#included`'ed so let's make it a header file.

File: linux.c [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  float uptime;
+
+  int uptime_fd;
+  char uptime_buf[100];
+  int uptime_buf_read;
+
+  uptime_fd = open("/proc/uptime", O_RDONLY);
```
Comment (kasperisager) [inline linux.c]: This should be `close()`'ed when no longer in use.

File: linux.c [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  sscanf(uptime_buf, "%f", &uptime);
+
+  DIR *self_task_dir;
+
+  char self_task_dir_path[35];
+  struct dirent *self_task_dir_entry;
+
+  sprintf(self_task_dir_path, "/proc/%d/task", self);
+
+  self_task_dir = opendir(self_task_dir_path);
```
Comment (kasperisager) [inline linux.c]: This should be `closedir()`'ed when no longer in use.

File: linux.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    int pid = atoi(self_task_dir_entry->d_name);
+
+    char stat_path[30];
+    int stat_fd;
+
+    sprintf(stat_path, "/proc/%d/task/%d/stat", self, pid);
+
+    stat_fd = open(stat_path, O_RDONLY);
```
Comment (kasperisager) [inline linux.c]: This should be `close()`'ed when no longer in use.

**Review by kasperisager — CHANGES_REQUESTED** [review]
Great first draft! In addition to the other points, we also write into quite a few buffers of varying sizes without really explaining and/or ensuring that the sizes are appropriate. You can probably get away with writing into a single buffer and then giving it a Sufficiently Large(tm) size of e.g. 4 or 8 KiB.

File: linux.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    return -1;
+  }
+
+  char uptime_buf[4096 + 1];
+  int uptime_buf_read;
+
+  uptime_buf_read = read(uptime_fd, uptime_buf, 4096);
+
+  if (uptime_buf_read == -1) {
+    return -1;
```
Comment (kasperisager) [inline linux.h]: If this is hit then `uptime_fd` leaks. It should be closed prior to checking the read status.

File: linux.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  err = close(uptime_fd);
+
+  if (err == -1) {
+    return -1;
+  }
```
Comment (kasperisager) [inline linux.h]:
```suggestion
  err = close(uptime_fd);
  assert(err == 0);
```

Closing valid file descriptors is one of those things that aren't allowed to fail in practice so it's safe to assert to it succeeds.

File: linux.h [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    if (stat_buf_read == -1) {
+      continue;
+    }
+
+    err = close(stat_fd);
+
+    if (err == -1) {
+      break;
+    }
```
Comment (kasperisager) [inline linux.h]:
```suggestion
    err = close(stat_fd);
    assert(err == 0);
```

Same as above.

File: linux.h [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      continue;
+    }
+
+    char stat_buf[4096 + 1];
+    int stat_buf_read;
+
+    stat_buf_read = read(stat_fd, stat_buf, 4096);
+
+    if (stat_buf_read == -1) {
+      continue;
```
Comment (kasperisager) [inline linux.h:70]: Same as above; this risks leaking `stat_fd`.

File: linux.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    stat->cpu_usage = (utime + stime) / (uptime - (starttime / clock_ticks));
+
+    ++threads_len;
+  }
+
+  err = closedir(self_task_dir);
+
+  if (err == -1) {
+    return -1;
+  }
```
Comment (kasperisager) [inline linux.h]:
```suggestion
  err = closedir(self_task_dir);
  assert(err == 0);
```

Same as above.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: linux.h (hunk 6 above)
Comment (yassernasc) [reply linux.h]: Well pointed, leaks fixed at the new commit.

**Review by kasperisager — APPROVED** [review]
Looks solid! I made some small adjustments as well, should be good to go 🚀

**Pull request merged**